### PR TITLE
feature: re-implement lua's function `os.getenv` with FFI

### PR DESCRIPTION
**Pull request by hongliang5316**
See https://github.com/openresty/lua-nginx-module/issues/1340 and https://github.com/openresty/lua-nginx-module/pull/1341
Signed-off-by: hongliang <513918845@qq.com>

I hereby granted the copyright of the changes in this pull request
to the authors of this lua-resty-core project.

### CODE REVIEW

File: .luacheckrc [hunk 1]
```diff
@@ -1,5 +1,5 @@
 std = 'ngx_lua'
-globals = { 'ngx' }
+globals = { 'ngx', 'os.getenv' }
```
Comment (thibaultcha) [inline .luacheckrc]: tip: we can avoid having to do this is we load the `os` module locally, as done in #198

Comment (hongliang5316) [reply .luacheckrc]: good tip

File: lib/resty/core/os.lua [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+local _M = { version = base.version }
+
+
+function os.getenv(varname)
+    local cycle = getfenv(0).__ngx_cycle
+    if not cycle then
+        error("no cycle found")
+    end
+
+    local r = getfenv(0).__ngx_req
```
[the remark this answers is not in the document]
Comment (hongliang5316) [reply lib/resty/core/os.lua]: good point, thanks

File: lib/resty/core/os.lua [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+local ffi = require "ffi"
+local C = ffi.C
+local ffi_str = ffi.string
+local getfenv = getfenv
+local error = error
+
+
+ffi.cdef[[
+
+struct ngx_cycle_s;
```
[the remark this answers is not in the document]
Comment (hongliang5316) [reply lib/resty/core/os.lua]: @spacewander  This kept style of `ngx_http_request_t` See https://github.com/openresty/lua-resty-core/blob/master/lib/resty/core/base.lua#L95

File: t/os.t [hunk 4]
```diff
@@ -0,0 +1,82 @@
...
+#master_process_enabled(1);
+#log_level('warn');
+
+repeat_each(2);
+
+plan tests => repeat_each() * (blocks() * 3);
+
+our $pwd = cwd();
+
+$ENV{MyTestSysEnv} = 'bar';
```
[the remark this answers is not in the document]
Comment (hongliang5316) [reply t/os.t]: This is used for both TEST
i want to test `env MyTestSysEnv ` and `env MyTestSysEnv=hello` cases.

[the remark this answers is not in the document]
Comment (spacewander) [reply t/os.t]: Got it.

File: lib/resty/core/os.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (spacewander) [reply lib/resty/core/os.lua]: @hongliang5316 
The `struct ngx_http_request_s` is removed now.

File: lib/resty/core/os.lua [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+
+    local cycle = t.__ngx_cycle
+    if not cycle then
+        error("no cycle found")
+    end
+
+    local r = t.__ngx_req
+
+    local value = C.ngx_http_lua_ffi_os_getenv(cycle, r, varname)
+    if value then
```
Comment (spacewander) [inline lib/resty/core/os.lua]: The check here is incorrect.
In Lua, only `false` and `nil` are falsy value, so `cdata:NULL` is a truth value.
Need to check it with `if value ~= nil`.
BTW, better to add a test case here.

Comment (hongliang5316) [reply lib/resty/core/os.lua]: @spacewander 
Okay, i have test this, in LuaJIT `cdata<char *>: NULL` is truth, we can use `if cdata == nil` to check if `cdata` is a NULL pointer, `ffi.string(nil)` and `ffi.string(cdata<char *>: NULL)` 
will result segmentation fault, thanks!

**Comment by thibaultcha** [comment]
This is now implemented in c193534071f4f2e8335839961728e51c7c248d6a, closing this PR, thanks.

**Pull request closed without merging**